### PR TITLE
Latching faults and warnings

### PR DESCRIPTION
**Pull request by gtaharaedmonds**
### Summary

Faults and warnings (finally) latch. This is implemented in another CAN message for faults and warnings, which counts the number of times an alert is set. Each count gets 3 bits for  a maximum count of 7. This also means the maximum number of faults or warnings is reduced from 64 to 21, since currently we just pack all faults/warnings counts into a single CAN message and so can only fit 21 counts for a total of 63 bits.

So:

- `<board>_Faults` is unchanged (still describes the current status of each fault
- `<board>_FaultCounts` allocates a 3-bit counter for each fault, which count the number of rising edge transitions (no fault -> fault)

(and similarly for warnings)

### Changelist 

- Implement fault/warning counts as described above and fix tests
- Change the alert settings syntax to be more similar to `App_CanTx`/`App_CanRx`

### Testing Done

- [x] Spun wheels on jacks

### Checklist
*Please change `[ ]` to `[x]` when you are ready.*
- [x] I have read and followed the code conventions detailed in [README.md](../README.md) (*This will save time for both you and the reviewer!*).
- [x] If this pull request is longer then **500** lines, I have provided *explicit* justification in the summary above explaining why I *cannot* break this up into multiple pull requests (*Small PR's are faster and less painful for everyone involved!*).

### CODE REVIEW

**Review by Lucien950 — CHANGES_REQUESTED** [review]
a few small changes requested, otherwise builds just fine on a rebase

**Comment by Lucien950** [comment]
I noticed that alerts also interface with the same tx functions that can messages use, meaning that it is possible to set alerts while bypassing `App_CanAlerts...`. Should we move the logic for alert table mutation inside `App_CanAlerts.c`?

File: scripts/code_generation/jsoncan/src/codegen/c_generation/app_can_alerts_module.py [hunk 1]
```diff
@@ -154,12 +132,12 @@ def _public_functions(self) -> List[CFunc]:
...
-        funcs.append(self._set_alert_func(CanAlertType.WARNING, "warning"))
-        funcs.append(self._set_alert_func(CanAlertType.FAULT, "fault"))
+        funcs.extend(self._set_alert_funcs(CanAlertType.WARNING))
+        funcs.extend(self._set_alert_funcs(CanAlertType.FAULT))
 
         # Alert getters
-        funcs.append(self._get_alert_func(CanAlertType.WARNING, "warning"))
-        funcs.append(self._get_alert_func(CanAlertType.FAULT, "fault"))
+        funcs.extend(self._get_alert_funcs(CanAlertType.WARNING))
+        funcs.extend(self._get_alert_funcs(CanAlertType.FAULT))
```
[the remark this answers is not in the document]
Comment (gtaharaedmonds) [reply scripts/code_generation/jsoncan/src/codegen/c_generation/app_can_alerts_module.py:140]: Yep exactly.

File: scripts/code_generation/jsoncan/src/json_parsing/json_can_parsing.py [hunk 2]
```diff
@@ -157,13 +162,18 @@ def _parse_json_data(self, dir: str):
...
-                    CanAlert(alert.name, CanAlertType.FAULT) for alert in faults.signals
+                    *[
+                        CanAlert(alert.name, CanAlertType.WARNING)
+                        for alert in warnings.signals
+                    ],
+                    *[
+                        CanAlert(alert.name, CanAlertType.FAULT)
+                        for alert in faults.signals
+                    ],
                 ]
```
[the remark this answers is not in the document]
Comment (gtaharaedmonds) [reply scripts/code_generation/jsoncan/src/json_parsing/json_can_parsing.py:177]: Spread operator is sexier

File: scripts/code_generation/jsoncan/src/json_parsing/schema_validation.py [hunk 3]
```diff
@@ -96,7 +96,9 @@
     Or(
         {
             "warnings_id": And(int, lambda x: x >= 0),
+            "warnings_counts_id": And(int, lambda x: x >= 0),
             "faults_id": And(int, lambda x: x >= 0),
+            "faults_counts_id": And(int, lambda x: x >= 0),
```
[the remark this answers is not in the document]
Comment (gtaharaedmonds) [reply scripts/code_generation/jsoncan/src/json_parsing/schema_validation.py:101]: good call, added

**Comment by gtaharaedmonds** [comment]
> I noticed that alerts also interface with the same tx functions that can messages use, meaning that it is possible to set alerts while bypassing `App_CanAlerts...`. Should we move the logic for alert table mutation inside `App_CanAlerts.c`?

Yeah this is a fair concern, in fact we are currently already doing this in one or two places (fixed in my last PR). The alert signals are parsed as regular CAN signals, so that they're in the DBC and the pack/unpack functions are generated (like the rest of the CAN signals). This makes them very tightly coupled with the rest of the CAN system, so uncoupling them will be hard. As such I don't want to do this right now since latching faults is very high priority, maybe make a Jira ticket to look into this later?

**Review by Lucien950 — APPROVED** [review]
i ain't really check `can_bus/json` and `firmware/thruna` like that but python is cooking

**Pull request merged**